Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -0,0 +1,392 @@
+ * @parameter frameSize frame constrains to render the text on.
+ *
+ * @return the created IDWriteTextLayout object.
+ */
+static ComPtr<IDWriteTextLayout> _createDWriteTextLayout(_CTTypesetter* ts, CFRange range, CGRect frameSize) {
+    ComPtr<IDWriteTextFormat> textFormat = _createDWriteTextFormat(ts, range);
+
+    // TODO::
+    // Iterate through all attributed string ranges and identify attributes so they can be used to -
+    //  - set indentation
+    //  - set font characteristics
+    //  - etc.
+    //  These can be done using the DWrite IDWriteTextFormat range property methods.
+    NSRange curRange = NSMakeRangeFromCF(range);
+    /*
+    NSDictionary* attribs = [ts->_attributedString attributesAtIndex:0 effectiveRange:&curRange];
+    UIFont* font = [attribs objectForKey:static_cast<NSString*>(kCTFontAttributeName)];
+    if (font == nil) {
+        font = [s_lazyUIFont systemFontOfSize:c_defaultSystemFontSize];
+    }
+    float fontSize = [font pointSize];
+    */
+
+    wchar_t* wcharString = reinterpret_cast<wchar_t*>(
+        const_cast<char*>([[ts->_string substringWithRange:curRange] cStringUsingEncoding:NSUTF16StringEncoding]));
+
+    // TODO::
+    // We need too support widthFunc semantic to be able to support NSLayout*. We could either change the API signature of this API or
+    // create a new API to support dynamic width support.
+    //
+    // float width = widthFunc(widthParam, 0, 0, fontSize);
+
+    // Get the direct write factory instance
+    ComPtr<IDWriteFactory> dwriteFactory = GetDWriteFactoryInstance();
+
+    ComPtr<IDWriteTextLayout> textLayout;
+    THROW_IF_FAILED(dwriteFactory->CreateTextLayout(
+        wcharString, ts->_charactersLen, textFormat.Get(), frameSize.size.width, frameSize.size.height, &textLayout));
+
+    return textLayout;
+}
+
+/**
+ * Custom IDWriteTextRenderer class that captures the glyph runs generated by DWrite for the given TextLayout and TextRenderer constraints.
+ */
+class CustomTargetTextRenderer : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteTextRenderer> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    CustomTargetTextRenderer();
+
+    HRESULT RuntimeClassInitialize();
+
+    HRESULT STDMETHODCALLTYPE DrawGlyphRun(_In_ void* clientDrawingContext,
+                                           _In_ float baselineOriginX,
+                                           _In_ float baselineOriginY,
+                                           DWRITE_MEASURING_MODE measuringMode,
+                                           _In_ DWRITE_GLYPH_RUN const* glyphRun,
+                                           _In_ DWRITE_GLYPH_RUN_DESCRIPTION const* glyphRunDescription,
+                                           _In_ IUnknown* clientDrawingEffects) throw() {
+        _DWriteGlyphRunDetails* glyphs = static_cast<_DWriteGlyphRunDetails*>(clientDrawingContext);
```

Comment:
static_cast is only useful where types are related.  void * isnt really related.

---
In reply to: [77875281](https://github.com/Microsoft/WinObjC/pull/950#discussion_r77875281) [](ancestors = 77875281,77756151)